engine/signer.go

```diff
@@ -0,0 +1,146 @@
+	pubKey        *cryptotypes.PubKey
+	address       *cryptotypes.Address
+	bech32Address *sdktypes.AccAddress
+}
+
+func NewSigner(appchainCtx *appchainClient.Context, privKey *ed25519.PrivKey) *signer {
+	return &signer{
+		appchainCtx: appchainCtx,
+		privKey:     privKey,
+	}
+}
+
+func (s *signer) PubKey() cryptotypes.PubKey {
+	if s.pubKey != nil {
+		return *s.pubKey
+	} else {
+		pubKey := s.privKey.PubKey()
+		s.pubKey = &pubKey
+		return s.PubKey()
+	}
+}
+
+func (s *signer) Address() cryptotypes.Address {
+	if s.address != nil {
+		return *s.address
+	} else {
+		address := s.privKey.PubKey().Address()
+		s.address = &address
+		return s.Address()
+	}
+}
+
+func (s *signer) AccAddress() sdktypes.AccAddress {
+	if s.bech32Address != nil {
+		return *s.bech32Address
+	} else {
+		bech32Addr := sdktypes.AccAddress(s.Address())
+		s.bech32Address = &bech32Addr
+		return s.AccAddress()
+	}
+}
+
+func (s *signer) Sign(tx *sdktx.Tx) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("panic during tx signing: %v", r)
+		}
+	}()
+
+	txConfig := s.appchainCtx.TxConfig
+	txBuilder := txConfig.NewTxBuilder()
+
+	acc, err := s.appchainCtx.AccountRetriever.GetAccount(*s.appchainCtx, s.AccAddress())
+	if err != nil {
+		return fmt.Errorf("get account: %v", err)
+	}
+
+	signerData := authsigning.SignerData{
+		ChainID:       s.appchainCtx.ChainID,
+		AccountNumber: acc.GetAccountNumber(),
+		Sequence:      acc.GetSequence(),
+		PubKey:        s.PubKey(),
+		Address:       acc.GetAddress().String(),
+	}
+
+	blankSig := signing.SignatureV2{
+		PubKey:   s.PubKey(),
+		Sequence: acc.GetSequence(),
+		Data: &signing.SingleSignatureData{
+			SignMode:  signing.SignMode_SIGN_MODE_DIRECT,
+			Signature: nil,
+		},
+	}
+
+	err = txBuilder.SetMsgs(tx.Body.Messages[0])
+	if err != nil {
+		return fmt.Errorf("set msgs: %v", err)
+	}
+	err = txBuilder.SetSignatures(blankSig)
+	if err != nil {
+		return fmt.Errorf("set signatures: %v", err)
+	}
+
+	sig, err := cosmostx.SignWithPrivKey(
+		context.Background(),
+		signing.SignMode_SIGN_MODE_DIRECT,
+		signerData,
+		txBuilder,
+		s.privKey,
+		txConfig,
+		acc.GetSequence(),
+	)
+	if err != nil {
+		return fmt.Errorf("sign with priv key: %v", err)
+	}
+	err = txBuilder.SetSignatures(sig)
+	if err != nil {
+		return fmt.Errorf("set signatures: %v", err)
+	}
+
+	pubKeyAny, err := codectypes.NewAnyWithValue(s.PubKey())
+	if err != nil {
+		return fmt.Errorf("new any with value: %v", err)
+	}
+
+	tx.AuthInfo = &sdktx.AuthInfo{
+		SignerInfos: []*sdktx.SignerInfo{
+			{
+				PublicKey: pubKeyAny,
+				ModeInfo: &sdktx.ModeInfo{
+					// Assuming you want single signer mode
+					Sum: &sdktx.ModeInfo_Single_{
+						Single: &sdktx.ModeInfo_Single{
+							Mode: signing.SignMode_SIGN_MODE_DIRECT,
+						},
+					},
+				},
+				Sequence: acc.GetSequence(),
+			},
+		},
+		Fee: &sdktx.Fee{},
+	}
+	tx.Signatures = [][]byte{sig.Data.(*signing.SingleSignatureData).Signature}
```

Comment:
Is this logic to set the auth info and signatures already handled above by the [txBuilder.SetSignatures](https://github.com/polymerdao/monomer/pull/123/files#diff-7873bc6b8c386aec42ec5b42d310a7e2afacfc9f1a1ca12e97abe4df06b7eafcR116) call? If so, could these lines be removed?

Reply:
There's a nuisance indirection happening here.

The `Sign` function here takes a tx and decorates it with the signature data.

The `txBuilder` instantiates its own internal tx and operates relative to that. `SetSignatures` operates on the data on that inernal tx - not the tx that we are decorating.

There might be some gains in clarity if `Sign` returned a tx.

Reply:
Probably not worth it. The `tx` in a `txBuilder` is `github.com/cosmos/cosmos-sdk/x/auth/signing.Tx`, but the struct we're interested in manipulating / constructing is `github.com/cosmos/cosmos-sdk/types/tx.Tx`.